src/dev-app/datepicker/datepicker-demo.ts

```diff
@@ -59,8 +59,8 @@ export class DatepickerDemo {
       return !(date.getFullYear() % 2) && Boolean(date.getMonth() % 2) && !(date.getDate() % 2);
     }
 
-  onDateInput = (e: MatDatepickerInputEvent<Date>) => this.lastDateInput = e.value;
-  onDateChange = (e: MatDatepickerInputEvent<Date>) => this.lastDateChange = e.value;
+  onDateInput = (e: MatDatepickerInputEvent<Date | null>) => this.lastDateInput = e.value;
```

Comment:
It seems like the `| null` in the generic is proliferating across the code. It would be nice if we could clean this up a bit, maybe by making certain properties type `D | null` instead of `D`	so that it doesn't have to be passed around everywhere in the generics

Reply:
We can't do that, because it would make the `selection` of the range model `DateRange<D | null> | null`, but it's guaranteed to always be defined.

Reply:
Here's a commit that I think makes things a little better https://github.com/angular/components/commit/ee3069d621532cb3b7430951cf7a8e832cf34fee
- Exports `ExtractDateTypeFromSelection` to use it in `MatDatepickerInputBase` instead of assuming `S = D`
- Uses the concrete selection model types in classes where we know which one we're dealing with (optional, makes the types read cleaner)
- Switches `MatDatepickerInputEvent` to be generic on `D` and optionally `S` (which is assumed to be `unknown` if unspecified). If people are using the target at all it's probably for equality check, so in 99% of cases I don't think the type of `S` matters

Reply:
Most of the changes make sense, but we can't use `MatSingleDateSelectionModel` and `MatRangeDateSelectionModel` in the constructor, because we use `MatDateSelectionModel` as the DI token.

Reply:
Ok, fine to revert those then

Reply:
The correct replacement types would be `MatDateSelectionModel<D | null, D>` and `MatDateSelectionModel<DateRange<D>, D>`

Reply:
I've pushed the changes.